### PR TITLE
feat: support some metrics being 'None' without stopping training

### LINKED ISSUE #165: bug: logging_steps greater than one results in TypeError when evaluating trainer controller rule
## Describe the bug

logging_steps greater than one results in TypeError when evaluating trainer controller rule. Following trainer controller configuration used when the bug was detected:
```
controller-metrics:
  - name: loss
    class: Loss
controllers:
  - name: loss-controller
    triggers:
      - on_log
    rule: loss < 1.0
    operations:
      - hfcontrols.should_training_stop
```

## Platform

- Interpreter version: We ran the `./tuning/sft_trainer.py` script with the below training configuration in CCC cluster with 1 GPU (Tesla V100-SXM2-32GB).
- Library version: This environment was not used.

## Sample Code
Below configuration for `fms-hf-tuning` causes the crash:
```
python ./tuning/sft_trainer.py  \
--model_name_or_path $MODEL_PATH  \
--tokenizer_name_or_path $MODEL_PATH \
--training_data_path $DATA_PATH  \
--output_dir $OUTPUT_PATH  \
--validation_data_path $VALIDATION_PATH \
--num_train_epochs 10  \
--per_device_train_batch_size 4  \
--per_device_eval_batch_size 4  \
--gradient_accumulation_steps 4  \
--evaluation_strategy "epoch"  \
--save_strategy "epoch"  \
--learning_rate 1e-5  \
--weight_decay 0.  \
--warmup_ratio 0.03  \
--lr_scheduler_type "cosine"  \
--logging_steps 5  \
--logging_strategy "steps" \
--include_tokens_per_second  \
--packing False  \
--metric_for_best_model "loss" \
--load_best_model_at_end True \
--use_flash_attn False \
--trainer_controller_config_file "examples/trainercontroller_configs/loss.yaml" \
--response_template "\n### Response:"  \
--dataset_text_field "output" 
```

## Expected behavior

The rule should evaluate to False and should not throw any exception.

## Observed behavior
```
  File "/seshapad/fms-hf-tuning/tuning/trainercontroller/callback.py", line 235, in _take_control_actions
    raise TypeError("Rule failed due to incorrect type usage") from et
TypeError: Rule failed due to incorrect type usage
```

## Additional context

None.

### PR DESCRIPTION
**Pull request by HarikrishnanBalagopal**
Some metrics may not be available at the time of rule evaluation. Add some more unit tests for the same conditions.

### Description of the change

Fixes #165 

### Related issue number

#165 

### How to verify the PR

`make test`

### Was the PR tested

- [x] I have added >=1 unit test(s) for every new method I have added.
- [x] I have ensured all unit tests pass

### CODE REVIEW

**Comment by kmehant** [comment]
@HarikrishnanBalagopal can you fix the lint issues?

File: tuning/utils/evaluator.py [hunk 1]
```diff
@@ -5,16 +5,35 @@
 from simpleeval import DEFAULT_FUNCTIONS, DEFAULT_NAMES, EvalWithCompoundTypes
 
 
-def get_evaluator(metrics: dict) -> EvalWithCompoundTypes:
+class MetricUnavailableError(Exception):
+    def __init__(self, name):
+        super().__init__(f"The metric '{name}' is not available")
+        self.name = name
```
Comment (kmehant) [inline tuning/utils/evaluator.py:11]: Do we need this?
```suggestion

```

**Comment by HarikrishnanBalagopal** [comment]
> @HarikrishnanBalagopal can you fix the lint issues?

Done, but locally I also get this import error when doing `make lint`

```
lint: commands[0]> pylint tuning 'scripts/*.py' 'build/*.py' tests
************* Module pull_and_format_datasets
scripts/pull_and_format_datasets.py:13:0: E0401: Unable to import 'boto3' (import-error)
```

I am able to import it in my local `venv`
```
$ python3
Python 3.11.5 (main, Aug 24 2023, 15:09:45) [Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import boto3
```

**Comment by HarikrishnanBalagopal** [comment]
Screenshot for Step 5
image

**Review by kmehant — APPROVED** [review]
PR adds an important feature for trainer controller. It also includes test cases for validation. The changes are mostly concentrated on the `trainercontroller` module. Existing test cases as well pass and does not modify user facing interface.

**Comment by seshapad** [comment]
Closes #183, #184, #185, #186 and #187.

**Pull request merged**